Maui Net6/7: Undefined Symbols for architecture arm64 error

When attempting to build both Maui net6 and net7 versions for iOS, clang hits a linker error. These appear to be the same on both the net6 and net7 versions. Similar build processes are used for both but will use the net7 version for the reproduction steps. This may be better off in the Maui repo, but figured to start here given the error is specifically coming from Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14 on for net7 and Microsoft.iOS.Sdk/16.0.515 for net6 with reference to Xamarin targets.

### Steps to Reproduce
#### General
1. Download latest dotnet 7
2. Install the maui workload ensuring to point to latest packages (https://aka.ms/dotnet/maui/net7.0.json)
3. Generate the boiler plate Maui Project
4. Restore and build the project

#### Specific (Exact steps given clean environment, does download files)
```
curl -o NuGet.config 'https://raw.githubusercontent.com/dotnet/maui/main/NuGet.config'
curl -o dotnet-install.sh 'https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.sh'
curl -Lo maui-supported-sdk-version.json 'https://aka.ms/dotnet/sdk/maui/net7.0.json'
chmod -R a+rx .
./dotnet-install.sh --version 7.0.100-rc.2.22477.20 --install-dir .
./dotnet --info
./dotnet workload install maui --from-rollback-file https://aka.ms/dotnet/maui/net7.0.json --configfile NuGet.config
./dotnet new maui -n MauiTesting
cd MauiTesting
cp $(Build.SourcesDirectory)/NuGet.config ./NuGet.config
# remove net7.0-maccatalyst to work around https://github.com/dotnet/sdk/issues/21877
cp MauiTesting.csproj MauiTesting.csproj.bak
sed -i'' -e 's/net7.0-ios;net7.0-maccatalyst/net7.0-ios/g' MauiTesting.csproj
../dotnet publish -bl:MauiiOS.binlog -f net7.0-ios --self-contained -r ios-arm64 -c Release /p:_RequireCodeSigning=false /p:ApplicationId=net.dot.mauitesting
```

### Expected Behavior
The app should build into an ipa file that can be run for testing.

### Actual Behavior
We get to the linker step and then fail.
```
  Optimizing assemblies for size may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Optimizing assemblies for size. This process might take a while.
ILLINK : warning MT0079: The recommended Xcode version for Microsoft.iOS 16.0.1263 is Xcode 14.0 or later. The current Xcode version (found in /Applications/Xcode_13.2.1.app/Contents/Developer) is 13.2.1. [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
  		
  Tool xcrun execution finished (exit code = 1).
          
  Undefined symbols for architecture arm64:
    "_objc_msgSend$valueWithUIOffset:", referenced from:
        _xamarin_uioffset_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o)
    "_objc_msgSend$valueWithSCNVector4:", referenced from:
        _xamarin_scnvector4_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o)
    "_objc_msgSend$valueWithSCNMatrix4:", referenced from:
        _xamarin_scnmatrix4_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o)
    "_objc_msgSend$valueWithMKCoordinateSpan:", referenced from:
        _xamarin_mkcoordinatespan_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o)
    "_objc_msgSend$valueWithMKCoordinate:", referenced from:
................................................. (Much of the same)
        _xamarin_release_managed_ref in libxamarin-dotnet.a(runtime-dotnet.arm64.o)
    "_objc_msgSend$numberWithUnsignedLongLong:", referenced from:
        _xamarin_ulong_to_nsnumber in libxamarin-dotnet.a(trampolines-dotnet.arm64.o)
    "_objc_msgSend$componentsJoinedByString:", referenced from:
        xamarin_compute_trusted_platform_assemblies() in libxamarin-dotnet.a(runtime-dotnet.arm64.o)
        xamarin_compute_native_dll_search_directories() in libxamarin-dotnet.a(runtime-dotnet.arm64.o)
    "_objc_msgSend$release", referenced from:
        _xamarin_printf in libxamarin-dotnet.a(runtime-dotnet.arm64.o)
        _xamarin_release_managed_ref in libxamarin-dotnet.a(runtime-dotnet.arm64.o)
        ___32-[XamarinNSThreadObject dealloc]_block_invoke in libxamarin-dotnet.a(shared-dotnet.arm64.o)
        _xamarin_log in libxamarin-dotnet.a(xamarin-support-dotnet.arm64.o)
        _xamarin_set_gchandle_trampoline in libxamarin-dotnet.a(trampolines-dotnet.arm64.o)
  ld: symbol(s) not found for architecture arm64
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
  
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error : clang++ exited with code 1: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error : Undefined symbols for architecture arm64: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :   "_objc_msgSend$valueWithUIOffset:", referenced from: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :       _xamarin_uioffset_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o) [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :   "_objc_msgSend$valueWithSCNVector4:", referenced from: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :       _xamarin_scnvector4_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o) [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :   "_objc_msgSend$valueWithSCNMatrix4:", referenced from: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :       _xamarin_scnmatrix4_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o) [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :   "_objc_msgSend$valueWithMKCoordinateSpan:", referenced from: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :       _xamarin_mkcoordinatespan_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o) [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :   "_objc_msgSend$valueWithMKCoordinate:", referenced from: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :       _xamarin_cllocationcoordinate2d_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o) [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :   "_objc_msgSend$valueWithCGVector:", referenced from: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :       _xamarin_cgvector_to_nsvalue in libxamarin-dotnet.a(trampolines-dotnet.arm64.o) [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :   "_objc_msgSend$valueWithCMTimeRange:", referenced from: [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
/Users/runner/work/1/s/packs/Microsoft.iOS.Sdk/16.0.1263-rc.2-xcode14/targets/Xamarin.Shared.Sdk.targets(1230,3): error :       _xamarin_cmtimerange_to_nsvalu [/Users/runner/work/1/s/MauiTesting/MauiTesting.csproj::TargetFramework=net7.0-ios]
```

### Environment
<details>
<summary>Runtime Information</summary>

```
.NET SDK:
 Version:   7.0.100-rc.2.22477.20
 Commit:    915e3f7110

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  11.7
 OS Platform: Darwin
 RID:         osx.11.0-x64
 Base Path:   /Users/runner/work/1/s/sdk/7.0.100-rc.2.22477.20/

Host:
  Version:      7.0.0-rc.2.22464.1
  Architecture: x64
  Commit:       c623d96fe3

.NET SDKs installed:
  7.0.100-rc.2.22477.20 [/Users/runner/work/1/s/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 7.0.0-rc.2.22464.13 [/Users/runner/work/1/s/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 7.0.0-rc.2.22464.1 [/Users/runner/work/1/s/shared/Microsoft.NETCore.App]
```
</details>
<details>
<summary>Workloads used in rollback</summary>

```
{
"microsoft.net.sdk.android": "33.0.0-rc.2.202/7.0.100-rc.2",
"microsoft.net.sdk.ios": "16.0.1263-rc.2-xcode14/7.0.100-rc.2",
"microsoft.net.sdk.maccatalyst": "15.4.1180-rc.2/7.0.100-rc.2",
"microsoft.net.sdk.macos": "12.3.1180-rc.2/7.0.100-rc.2",
"microsoft.net.sdk.maui": "7.0.0-rc.2.6802/7.0.100-rc.2",
"microsoft.net.sdk.tvos": "16.0.1263-rc.2-xcode14/7.0.100-rc.2",
"microsoft.net.workload.mono.toolchain.net6": "7.0.0-rc.2.22472.3/7.0.100",
"microsoft.net.workload.mono.toolchain.net7": "7.0.0-rc.2.22472.3/7.0.100",
"microsoft.net.workload.emscripten.net6": "7.0.0-rc.2.22465.1/7.0.100",
"microsoft.net.workload.emscripten.net7": "7.0.0-rc.2.22465.1/7.0.100"
}
```
</details>

### Build Logs
[MauiiOS.zip](https://github.com/xamarin/xamarin-macios/files/9701619/MauiiOS.zip)

--- a/Make.config
+++ b/Make.config
@@ -355,7 +355,11 @@ endif
 
 SIMULATOR_SDK=$(XCODE_DEVELOPER_ROOT)/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator$(IOS_SDK_VERSION).sdk
 
-OBJC_CFLAGS=-ObjC++ -std=c++14 -fno-exceptions -stdlib=libc++
+# Clang will by default emit objc_msgSend stubs in Xcode 14, which ld from earlier Xcodes doesn't understand.
+# We disable this by passing -fno-objc-msgsend-selector-stubs to clang.
+# We can probably remove this flag once we require developers to use Xcode 14.
+# Ref: https://github.com/xamarin/xamarin-macios/issues/16223
+OBJC_CFLAGS=-ObjC++ -std=c++14 -fno-exceptions -stdlib=libc++ -fno-objc-msgsend-selector-stubs
 
 COMMON_SIMULATOR_CFLAGS=-mios-simulator-version-min=$(MIN_IOS_SDK_VERSION) -isysroot $(SIMULATOR_SDK) $(CFLAGS) -g $(IOS_COMMON_DEFINES)
 SIMULATOR86_CFLAGS=$(COMMON_SIMULATOR_CFLAGS) -arch i386

